SimplePaymentsDialog: move 'activeTab' to component state

jsnajdr opened the pull request:
Move the `activeTab` property to the SimplePaymentsDialog component state, instead of
maintaining it outside and passing it down as props.

We'll need to switch tabs internally when implementing the "Edit" action -- in that case,
we'll switch from the 'list' tab to the 'form' and back.

When the dialog is being shown, ensure that it always shows the 'form' immediately after
being opened by implementing a check in `componentWillReceiveProps`.

I also renamed the `activeTab` values from `paymentButtons`/`addNew` to `list`/`form`. It's
shorter, and we'll be showing the `form` not only when adding a new button, but also when editing
a new one.

retrofox commented:
@jsnajdr could you add some testing instructions please?

jsnajdr commented:
Oh, sorry @retrofox -- I keep forgetting about testing plans :(

In short, everything should work as before without any observable changes :)

Workflows that should be checked if they didn't break:
- open the payments dialog, check that the "Add New" form is displayed. Fill something and then click "Cancel". After opening the dialog again, the form should be empty.
- switch to the "Payment Buttons" list, then click "Cancel". Open the dialog again. We should be at the "Add New" form now -- the displayed tab was changed when dialog was being opened.

retrofox commented:
> Oh, sorry @retrofox -- I keep forgetting about testing plans :(

I'm sorry I've been a little bit away from this side. thanks!

retrofox commented on client/components/tinymce/plugins/simple-payments/dialog/index.jsx:
```diff
@@ -62,13 +60,21 @@ class SimplePaymentsDialog extends Component {
 		} );
 
 		this.state = {
```
Just FYI you could set the initial state out of the constructor.

```es6
class SimplePaymentsDialog extends Component {
  state = {
    activeTab: 'form',
    selectedPaymentId: null,
    form: this.formStateController.getInitialState(),
    isSubmitting: false,
    errorMessage: null,
  }
```

jsnajdr replied:
I can't do this if the state depends on `this.formStateController`. See how Babel transpiles the code that uses class properties (tested on http://babeljs.io/repl):

Source:
```js
class Test extends Component {
  state = {
    activeTab: this.props.activeTab,
    form: this.controller.form()
  };
  constructor(props) {
    super(props);
    this.controller = new Controller();
  }
}
```
Transpiled:
```js
class Test extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activeTab: this.props.activeTab,
      form: this.controller.form()
    };
    this.controller = new Controller();
  }
}
```
`super` is called first, that's why I can use `this.props` when assigning to state. The rest of the constructor, however, is called only after the property assignments.

retrofox replied:
Thanks for the explanation Jarda.

retrofox commented on client/components/tinymce/plugins/simple-payments/dialog/index.jsx:
```diff
@@ -107,6 +113,10 @@ class SimplePaymentsDialog extends Component {
 		onComplete( null, formErrors );
 	}
 
+	handleChangeTabs = activeTab => {
```
maybe we could define this method in only one line:

```es6
handleChangeTabs = activeTab => this.setState( { activeTab } );
```

retrofox commented on client/components/tinymce/plugins/simple-payments/dialog/index.jsx:
```diff
@@ -107,6 +113,10 @@ class SimplePaymentsDialog extends Component {
+		this.setState( { activeTab } );
+	};
+
 	handleSelectedChange = selectedPaymentId => {
```
I'm aware that it isn't the primary purpose of this PR but we could do the same here that above.

jsnajdr commented:
- rebased the PR on top of latest `master`
- converted simple methods to one-line arrow functions
- added a `this._isMounted` guard to `showError` and `dismissError`

retrofox commented:
It works as expected just for a little detail. I'm not sure that it was introduced on this patch though.
If yow go to `Payment Buttons` and then press the `ESC` key the dilaog isn't closed. You need to click inside of the dialog in order to close it pressing the `ESC` key.

Beyond of this it wotks really good.

jsnajdr commented:
> If yow go to Payment Buttons and then press the ESC key the dilaog isn't closed. You need to click inside of the dialog in order to close it pressing the ESC key.

I verified that this happens even when this patch is not applied. Most likely some problem with the modal dialog element not having focus.
